File: src/differential_equations_first_order_linear_easy.py
"""
differential_equations - first_order_linear (easy)
Generated: 2026-02-22T05:20:01.222580
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # dy/dx + a*y = 0  (homogeneous)
        a = nonzero(-6, 6)

        C = symbols('C')
        ans = C * exp(-a * x)

        return problem(
            question=f"Solve the differential equation: $\\frac{{dy}}{{dx}} + {a}y = 0$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="differential_equations/first_order_linear",
            solution=steps(
                f"Separate: $\\frac{{dy}}{{y}} = -{a}\\,dx$",
                f"Integrate: $\\ln|y| = {-a}x + C_1$",
                f"$y = Ce^{{{-a}x}}$"
            ),
        )

    elif problem_type == 2:
        # dy/dx + a*y = b  (constant forcing)
        a = nonzero(-5, 5)
        b = nonzero(-8, 8)

        C = symbols('C')
        ans = Rational(b, a) + C * exp(-a * x)

        return problem(
            question=f"Solve the differential equation: $\\frac{{dy}}{{dx}} + {a}y = {b}$",
            answer=ans,
            difficulty=(1100, 1200),
            topic="differential_equations/first_order_linear",
            solution=steps(
                f"Integrating factor: $\\mu = e^{{{a}x}}$",
                f"Multiply: $\\frac{{d}}{{dx}}[e^{{{a}x}}y] = {b}e^{{{a}x}}$",
                f"Integrate: $e^{{{a}x}}y = \\frac{{{b}}}{{{a}}}e^{{{a}x}} + C$",
                f"$y = {latex(Rational(b,a))} + Ce^{{{-a}x}}$"
            ),
        )

    elif problem_type == 3:
        # dy/dx = k*y  (growth/decay, wider range)
        k = nonzero(-6, 6)

        C = symbols('C')
        ans = C * exp(k * x)

        return problem(
            question=f"Solve the differential equation: $\\frac{{dy}}{{dx}} = {k}y$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="differential_equations/first_order_linear",
            solution=steps(
                f"Separate: $\\frac{{dy}}{{y}} = {k}\\,dx$",
                f"Integrate: $\\ln|y| = {k}x + C_1$",
                f"$y = Ce^{{{k}x}}$"
            ),
        )

    elif problem_type == 4:
        # dy/dx + y = a*x + b  (linear forcing)
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)

        # IF: e^x; d/dx[e^x*y] = (a*x+b)*e^x
        # Integrate by parts: e^x*(a*x+b) int = (a*x + b - a)*e^x + C  (for a*x)
        # y_p = a*x + b - a = a*(x-1) + b
        C = symbols('C')
        y_p = a * x + b - a
        ans = y_p + C * exp(-x)

        return problem(
            question=f"Solve the differential equation: $\\frac{{dy}}{{dx}} + y = {latex(a*x+b)}$",
            answer=ans,
            difficulty=(1200, 1280),
            topic="differential_equations/first_order_linear",
            solution=steps(
                f"Integrating factor: $\\mu = e^x$",
                f"$\\frac{{d}}{{dx}}[e^x y] = ({latex(a*x+b)})e^x$",
                f"Integrate by parts: $e^x y = ({latex(y_p)})e^x + C$",
                f"$y = {latex(y_p)} + Ce^{{-x}}$"
            ),
        )

    else:
        # dy/dx + a*y = b*e^(c*x) with c != -a
        a = nonzero(-4, 4)
        b = nonzero(-5, 5)
        c = nonzero(-4, 4)
        while c == -a:
            c = nonzero(-4, 4)

        # y_p = b/(a+c) * e^(c*x)
        C = symbols('C')
        y_p = Rational(b, a + c) * exp(c * x)
        ans = y_p + C * exp(-a * x)

        return problem(
            question=f"Solve the differential equation: $\\frac{{dy}}{{dx}} + {a}y = {b}e^{{{c}x}}$",
            answer=ans,
            difficulty=(1150, 1270),
            topic="differential_equations/first_order_linear",
            solution=steps(
                f"Integrating factor: $\\mu = e^{{{a}x}}$",
                f"$\\frac{{d}}{{dx}}[e^{{{a}x}}y] = {b}e^{{{a+c}x}}$",
                f"Integrate: $e^{{{a}x}}y = \\frac{{{b}}}{{{a+c}}}e^{{{a+c}x}} + C$",
                f"$y = {latex(y_p)} + Ce^{{{-a}x}}$"
            ),
        )


emit(generate())

File: src/differential_equations_first_order_linear_easy.jl
# differential_equations - first_order_linear (easy)
# Generated: 2026-03-08T20:50:18.207112

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("differential_equations/first_order_linear")
    
    problem_type = rand(1:4)
    
    if problem_type == 1
        # Simple separable dy/dx = ky form (700-900 ELO)
        k = nonzero(-8, 8)
        C = randint(-15, 15)
        
        problem(
            question="Solve the differential equation: \$\\frac{dy}{dx} = $(k)y\$",
            answer="y = Ce^($(k)x)",
            difficulty=(700, 900),
            solution=steps(
                sol("Given", "\\frac{dy}{dx} = $(k)y"),
                "Separate variables: \$\\frac{dy}{y} = $(k)dx\$",
                "Integrate both sides: \$\\ln|y| = $(k)x + C\$",
                sol("General solution", "y = Ce^{$(k)x}"),
            ),
            time=90,
            answer_type="expression",
        )
        
    elseif problem_type == 2
        # dy/dx + P(x)y = 0 form (800-1000 ELO)
        p = nonzero(-6, 6)
        
        problem(
            question="Solve the differential equation: \$\\frac{dy}{dx} + $(p)y = 0\$",
            answer="y = Ce^($(-p)x)",
            difficulty=(800, 1000),
            solution=steps(
                sol("Given", "\\frac{dy}{dx} + $(p)y = 0"),
                "Rewrite as: \$\\frac{dy}{dx} = $(-p)y\$",
                "Separate and integrate: \$\\ln|y| = $(-p)x + C\$",
                sol("General solution", "y = Ce^{$(-p)x}"),
            ),
            time=100,
            answer_type="expression",
        )
        
    elseif problem_type == 3
        # First order linear with constant coefficients dy/dx + ay = b (900-1100 ELO)
        a = nonzero(-5, 5)
        b = nonzero(-12, 12)
        particular = b // (-a)
        
        problem(
            question="Find the general solution: \$\\frac{dy}{dx} + $(a)y = $(b)\$",
            answer="y = Ce^($(-a)x) + $(particular)",
            difficulty=(900, 1100),
            solution=steps(
                sol("Given", "\\frac{dy}{dx} + $(a)y = $(b)"),
                "Homogeneous solution: \$y_h = Ce^{$(-a)x}\$",
                "Particular solution (constant): \$y_p = $(particular)\$",
                sol("General solution", "y = Ce^{$(-a)x} + $(particular)"),
            ),
            time=120,
            answer_type="expression",
        )
        
    else
        # Initial value problem dy/dx = ky, y(0) = y0 (1000-1200 ELO)
        k = nonzero(-6, 6)
        y0 = nonzero(-10, 10)
        
        problem(
            question="Solve the initial value problem: \$\\frac{dy}{dx} = $(k)y\$, \$y(0) = $(y0)\$",
            answer="y = $(y0)e^($(k)x)",
            difficulty=(1000, 1200),
            solution=steps(
                sol("Given", "\\frac{dy}{dx} = $(k)y, \\quad y(0) = $(y0)"),
                "General solution: \$y = Ce^{$(k)x}\$",
                "Apply initial condition: \$$(y0) = Ce^{0} = C\$",
                sol("Particular solution", "y = $(y0)e^{$(k)x}"),
            ),
            time=120,
            answer_type="expression",
        )
    end
end